include/Mat.h

```diff
@@ -237,31 +253,37 @@ class Mat {
             throw invalid_argument("Initializer list size inconsistent with dimensions");
         buildStrides();
      
-        memory = new Type[size()];
-        data = memory;
+        data = new Type[size()];
         size_type i = 0;
         for(auto elem : list){
             data[i] = elem;
             i++;
         }
+        base = newMat(data);
     }
 
     Mat(const Mat& b){
-        refCount = b.refCount;
-        (*refCount)++;
+        if(b.base != NULL) base = b.base;
+        else base = const_cast<Mat<Type>*>(&b);
+        (base->refCount)++;
+
         ndim = b.ndim;
-        dims = new size_type[ndim];
-        for(long i = 0; i < ndim; i++){
-            dims[i] = b.dims[i];
+        data = b.data;
+        
+        if(b.allocator != NULL){
+            DimInfo dInfo(ndim, dims, strides);
```

Comment:
for now, assert allocator==NULL, and then we can write this part when we know how allocators actually work